Make H-loop-fusion share operands with users.

trentlo opened the pull request.

trentlo committed:
Make H-loop-fusion share operands with users.

trentlo commented:
@cheshire please help to take a look.

@akuegel @sanjoy FYI.

akuegel commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,135 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+      VLOG(3) << "Visiting: " << user->ToString();
+      switch (user->opcode()) {
+        case HloOpcode::kTuple:
+        // We say reshape is fine because it does not reorder elements.
```
There is the ReshapeIsBitcast function which you could use here to verify this.
Theoretically Reshape could reorder elements, but we usually try to assign a layout so that it doesn't.

trentlo replied:
Since we check row-major layout below, it is safe to assume reshape does not re-order elements. However, ShapeBitcast is better. Thanks for the pointer.

trentlo commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,135 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+          break;
+      }  // end of switch
+      if (user->opcode() != HloOpcode::kTuple &&
+          !LayoutUtil::IsMonotonicWithDim0Major(user->shape().layout())) {
```
This is a more restricted form. So long as all layouts are row major, we ensure no layout change. The HorizontalLoopFusion only supports row-major layouts. However, since this creates confusion, I will change to what you suggest.

trentlo commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,135 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+    return false;
+  }
+  int64 prev_limit = 0;
+  for (size_t i = 0; i < users.size(); ++i) {
```
Changed. Thanks for the catch!

trentlo reviewed (commented):
@akuegel thanks for the review. Please help to take a look again.

trentlo committed:
Revision to address review comments.

cheshire commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,139 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+         1 == instr->slice_strides().at(0);
+}
+
+bool IsSliceInputFusion(const HloInstruction& unnested_hlo) {
```
What do you think about rewriting this function using HLO matchers? (cf. pattern_matcher.h and the files which use it), it seems like it could be a lot smaller and a lot more readable?

cheshire commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,139 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+  bool all_1d_slices = absl::c_all_of(users, [](const HloInstruction* i) {
+    return Is1dSliceWithoutStrides(i);
+  });
+  if (!all_1d_slices) {
```
Very minor nit: you could probably inline?

`if (c_all_of(users, &Is1dSliceWithoutStrides)) { ... }` ?

cheshire commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,139 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+  });
+}
+
+bool ConcatHasNoEffect(const HloInstruction* concat) {
```
Question: what does the codegen look like in this case? Is it also able to figure out it's a noop, fuse it all, and don't emit anything extra?

trentlo replied:
In the case of HorizontalLoopFusion, the concat is inserted intentionally to increase the parallelism degree (by combining several fusions). So the concat will still be generated in the codegen.

The concat is simply generated by the elemental emitter.

cheshire reviewed (commented):
Impressive, thanks a lot!

trentlo committed:
[NFC] Polish coding style.

trentlo reviewed (commented):
Responded. Please take a look to see if it addresses all of your comments.

trentlo commented:
@cheshire 

I thought more about the problem and, unfortunately, I think the current solution (as-it) is not correct. Mainly, the concatenate can create create extra dependencies between its operands. Below is an example.

`    HloModule test

    fused_computation {
      // p0 has multiple transitive uses fed to concat and cannot share buffer
      // with outputs, because the aliased output could be written before all
      // the uses of p0 are finished.
      p0 = f32[100] parameter(0)
      p1 = f32[100] parameter(1)
      add0 = f32[100] add(p0, p1)
      concat0 = f32[200] concatenate(p0, add0), dimensions={0}
      slice0 = f32[100] slice(concat0), slice={[0:100]}
      slice1 = f32[100] slice(concat0), slice={[100:200]}
      ROOT tuple = (f32[100], f32[100]) tuple(slice0, slice1)
    }

    ENTRY test {
      p0 = f32[100] parameter(0)
      p1 = f32[100] parameter(1)
      ROOT fusion = (f32[100], f32[100]) fusion(p0, p1), kind=kInput, calls=fused_computation
    }
`

We can still follow the same way to prove that the transitive use closure is fed into only 1 concat and eventually into the user (that we want to set up alias.) Then, since the entire closure (inc. the user) shares the same iteration space (computed by the same thread ids), it is safe to share buffers.

I don't have your cases at hands and I'm a bit worried that given the new constraints it may not address your cases in concern. What do you feel about this? If there is a good chance to solve your cases, I'd say let's still give it a try.

Alternatively, we might have to step back to remove the concat from the HorizontalLoopFusion (i.e., fuse by sequentially running each fusion) and it might lose some performance in some cases.

I will also discuss internally within NVIDIA to get some feedback tomorrow to see what I should do. (@nouiz and @bas-aarts )

trentlo committed:
Fixing a bug that concat can create extra constraints to buffer sharing.

- Also, add a negative test.

trentlo committed:
Add one more unitest for h-loop-fusion buffer sharing.

cheshire commented:
Unfortunately though the benchmark which was previously OOMing is still OOMing with this patch =(

trentlo commented:
> Unfortunately though the benchmark which was previously OOMing is still OOMing with this patch =(

Ah~ That is indeed unfortunate. Does the OOM disappear with the original commit without the new constraints? I.e., https://github.com/tensorflow/tensorflow/pull/46614/commits/4edc3da19631fa3cb3525a2e4362262685d87e20

If it is the new constraints making it not work, we can simply exclude the cases with the constraints in the HorizontalLoopFusion. That is, we can simply exclude the cases where fusions share the same input parameters (because these input parameters will be used as different concat operands after the fusions are fused). Since this PR can already deal with the cases without sharing parameters across fusions, adding the exclusion should make the solution complete.

trentlo committed:
[XLA/GPU] Do not horizontally fuse loop-fusions if they share parameters.

trentlo commented:
> > Unfortunately though the benchmark which was previously OOMing is still OOMing with this patch =(
> 
> Ah~ That is indeed unfortunate. Does the OOM disappear with the original commit without the new constraints? I.e., [4edc3da](https://github.com/tensorflow/tensorflow/commit/4edc3da19631fa3cb3525a2e4362262685d87e20)
> 
> If it is the new constraints making it not work, we can simply exclude the cases with the constraints in the HorizontalLoopFusion. That is, we can simply exclude the cases where fusions share the same input parameters (because these input parameters will be used as different concat operands after the fusions are fused). Since this PR can already deal with the cases without sharing parameters across fusions, adding the exclusion should make the solution complete.

@cheshire, I just pushed a commit to disable fusing fusions that share parameters. Could you test if this solves your case? It should according to our understanding. (Otherwise, we have missed something.)

nouiz commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,172 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+    prev_concat_opnd_idx->first = &concat;
+    prev_concat_opnd_idx->second = operand_idx;
+  } else {
+    bool is_concat_identical = prev_concat_opnd_idx->first->Identical(
```
Also, why do you need the identical check instead of checking if they are exactly the same object? I think doing the same object check is simpler and cover what we need. If one input feed 2 concat, then it is a case we can't support or do not need.

trentlo replied:
(Unfortunately, ) the same transitive use closure can be fed to multiple concats.

Note that one concat is inserted to each output of the fusion candidates. I added some more descriptions and hopefully it is a bit clearer.

cheshire commented:
@trentlo I've shared the problematic HLOs with you, do you think you could take a look?

trentlo commented:
> @trentlo I've shared the problematic HLOs with you, do you think you could take a look?

Sure. Will do. Thanks for preparing the HLOs.

gbaned commented:
@trentlo  Any update on this PR? Please. Thanks!

trentlo commented:
> @trentlo I've shared the problematic HLOs with you, do you think you could take a look?

@cheshire, so, this PR works well for the HLOs you gave me although Adrian later traced the OOM issue under investigation to another major source.

Some details on the analysis of the HLOs you gave me:
1.	HorizontalLoopFusion is only active on module 0000 and 0157 (out of 8 modules you gave me.)
2.	Each module has only 1 horizontal_loop_fusion produced. The case in module 0000 falls into the case of sharing parameters between (fused) fusions, so the i/o buffers of the horizontal fusion are not sharable. However, this case is excluded by this commit (https://github.com/tensorflow/tensorflow/pull/46614/commits/24c07a32c800c0ab4d2705491a2c48c8c9bdbff8), which skips the fusions that share input parameters. So, the case is solved by the commit.
3.	The other case in Module 0157 can share i/o buffers. I don’t have your i/o aliasing setups, so I tried several (random) setups of i/o aliasing and the PR succeeds proving the share-ability (so long as the i/o buffers indeed have in-place update relationship)!

I'd suggest continuing the review process and merging the PR as we believe it addresses the io aliasing issue (that we've been discussing). Please let me know what you think. Thanks!

cheshire commented:
@trentlo Thanks a lot! Does it mean that we can remove the guard at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc#L181-L183 and that the test case at https://gist.github.com/cheshire/0b9683c2c84d096b95bbbf02958173f4 now does not create copies?

trentlo committed:
[XLA/GPU] Use IsElementwiseOnOperand() to test elementwise relationship.

Otherwise, we will exclude DynamicSliceUpdate because it is not a
pure elementwise instruction.

trentlo committed:
[XLA/GPU] Horizontally fuse DynamicUpdateSlice.

Previously we encountered an issue that additional copies are inserted due to
horizontal fusion. Re-enable the handling of DynamicUpdateSlice since a fix
has been committed.

With the fix, we observe a 30% speedup due to horizontal loop fusion from the
previously problematic case. See the below trace from nvprof.

//Horizontal fusion (w/ fix), kernel time ~= 3.33ms (30% speedup vs. w/o horizontal fusion.)
    29.51%  1.8112ms      1000  1.8110us  1.7910us  2.3360us  fusion_5
    24.68%  1.5148ms      1000  1.5140us  1.5030us  1.9840us  fusion_3
     0.09%  5.5040us         3  1.8340us  1.5360us  2.1120us  [CUDA memcpy DtoD]

//Horizontal fusion (w/o fix), kernel time ~= 5.74ms
    28.91%  2.4695ms      2003  1.2320us  1.2150us  1.9200us  [CUDA memcpy DtoD]
    21.13%  1.8052ms      1000  1.8050us  1.7910us  2.3360us  fusion_5
    17.20%  1.4691ms      1000  1.4690us  1.4390us  1.6320us  fusion_3

// No horizontal fusion, kernel time = ~4.36ms
    21.21%  1.5209ms      1000  1.5200us  1.4720us  10.110us  fusion_3
    19.81%  1.4204ms      1000  1.4200us  1.3750us  8.5750us  fusion_1
    19.74%  1.4155ms      1000  1.4150us  1.3750us  1.7600us  fusion
    0.07%  5.2800us         3  1.7600us  1.5040us  1.8880us  [CUDA memcpy DtoD]

trentlo commented:
> @trentlo Thanks a lot! Does it mean that we can remove the guard at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc#L181-L183 and that the test case at https://gist.github.com/cheshire/0b9683c2c84d096b95bbbf02958173f4 now does not create copies?

@cheshire thanks for the testcase!

This PR works for the test case after fixing a minor issue. I was not aware that DynamicUpdateSlice is not an elementwise op and we need to use IsElementwiseOnOperand() to test whether the compute is elementwise. The issue is fixed by this [commit](https://github.com/tensorflow/tensorflow/pull/46614/commits/3fb0103a881da1fdde23d2f3c2fde3dde2d3a988).

Out of curiosity, I measure the kernel time for benchmark using nvprof. Horizontal (loop) fusion now speeds up the case by 30%! See below for the relevant stats from nvprof.

```
//Horizontal fusion (w/ PR), kernel time ~= 3.33ms (30% speedup vs. w/o horizontal fusion.)
    29.51%  1.8112ms      1000  1.8110us  1.7910us  2.3360us  fusion_5
    24.68%  1.5148ms      1000  1.5140us  1.5030us  1.9840us  fusion_3
     0.09%  5.5040us         3  1.8340us  1.5360us  2.1120us  [CUDA memcpy DtoD]

//Horizontal fusion (w/o PR), kernel time ~= 5.74ms
    28.91%  2.4695ms      2003  1.2320us  1.2150us  1.9200us  [CUDA memcpy DtoD]
    21.13%  1.8052ms      1000  1.8050us  1.7910us  2.3360us  fusion_5
    17.20%  1.4691ms      1000  1.4690us  1.4390us  1.6320us  fusion_3

// No horizontal fusion, kernel time = ~4.36ms
    21.21%  1.5209ms      1000  1.5200us  1.4720us  10.110us  fusion_3
    19.81%  1.4204ms      1000  1.4200us  1.3750us  8.5750us  fusion_1
    19.74%  1.4155ms      1000  1.4150us  1.3750us  1.7600us  fusion
    0.07%  5.2800us         3  1.7600us  1.5040us  1.8880us  [CUDA memcpy DtoD]
```

trentlo committed:
Merge branch 'github-master' into h-fusion-sharing-opnd-with-user-upstreaming

trentlo commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,172 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+    return a->slice_starts().at(0) < b->slice_starts().at(0);
+  });
+  // Verify that each operand to the concat is reversed by a slice.
+  if (users.size() != concat.operand_count() ||
```
Done.

trentlo commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,172 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+    prev_limit = u->slice_limits().at(0);
+  }
+
+  // If we have seen other concats, make sure they are identical.
```
Added some more descriptions.

trentlo committed:
[XLA/GPU] Comment polishing.

trentlo commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,172 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+}
+
+// Returns whether we can prove the transitive uses of `param` are in effect
+// elementwise. In other words, we prove that the transitive use closure will
```
Done.

trentlo reviewed (commented):
@nouiz I addressed your comments. Thanks for the comments.

cheshire commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,171 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+// slices combinedly is a reverse function of the concat.
+// prev_concat_opnd_idx: previously seen concat and its operand index.
+//     prev_concat_opnd_idx.first is nullptr if no previously seen concat.
+bool ConcatIsEffectivelyElementwise(
```
Style nit: return value is preferred over output parameters.

Can we put the inout values in a struct? That would also give a better chance to explain their semantics.

E.g. something like

```
struct ConcatUsageInfo {
  const HloInstruction* concat;
  int64 op_idx;
  const HloInstruction* slice;
};

absl::optional<ConcatUsageInfo> ConcatIsEffectivelyElementwise(..., const ConcatUsageInfo& info);
```

?

trentlo replied:
Revised accordingly. Thanks for the suggestion.

cheshire commented on tensorflow/compiler/xla/service/hlo_dataflow_analysis.cc:
```diff
@@ -120,6 +120,171 @@ bool HloDataflowAnalysis::AreTransitiveUsesElementwiseOrTuple(
+  //   Slice Slice
+  //
+  std::vector<HloInstruction*> users = concat.users();
+  if (!absl::c_all_of(users, [](const HloInstruction* i) {
```
Nit: don't even need a lambda, can just directly put a function there (though up to you).

trentlo replied:
Nice catch.

cheshire commented:
> Out of curiosity, I measure the kernel time for benchmark using nvprof. Horizontal (loop) fusion now speeds up the case by 30%! 

Amazing ! BTW is there a short guide for using nvprof to get these numbers? I always kinda struggle when using the profiler.

trentlo committed:
[XLA/GPU] Polish coding styles.

trentlo commented on tensorflow/compiler/xla/service/copy_insertion_test.cc:
```diff
@@ -2877,6 +2877,58 @@ ENTRY main {
+  TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> module,
+                          ParseAndReturnVerifiedModule(hlo_string));
+  // Set up the aliasing manually which normally would be set by
+  // alias_passthrough_params pass.
```
Removed.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:
```diff
@@ -203,6 +195,24 @@ bool HasOnlyRowMajorLayout(const HloInstruction& fusion_instr) {
+bool AnyOpndIsParamSharedAmongFusions(
+    const HloInstruction* instr,
+    const absl::flat_hash_set<HloInstruction*>& fusion_instrs) {
+  for (const HloInstruction* opnd : instr->operands()) {
```
Revised. That is neat!

trentlo reviewed (commented):
Addressed the review comments. Please help to take a look again.

trentlo commented:
> > Out of curiosity, I measure the kernel time for benchmark using nvprof. Horizontal (loop) fusion now speeds up the case by 30%!
> 
> Amazing ! BTW is there a short guide for using nvprof to get these numbers? I always kinda struggle when using the profiler.

I usually just prepend `nvprof` to the command you want to run. For example, to replay the hlo, the following is what I did.

`nvprof ./bazel-bin/tensorflow/compiler/xla/tools/replay_computation_gpu --use_fake_data=true --print_result=false --compile_only=false dynamic_slice.hlo`

You will see (reasonably nice) summary of stats after the run.

For short run time, e.g., within few minutes, my experience with nvprof is good. I avoid using nvprof for long runtime because its post-processing time is too long for me. For longer run time, I use nsight system.

cheshire commented:
Thanks!

cheshire commented:
@trentlo Unfortunately, this change got rolled back. It crashes the test in `https://cs.opensource.google/jax/jax/+/master:tests/jet_test.py;l=297?q=JetTest.test_cumsum&sq=`

with

```
F0212 16:11:17.079444    7826 shape_util.cc:536] Check failed: shape.IsArray() (f64[6], f64[6], f64[6], f64[6])
*** Check failure stack trace: ***
    @     0x55c202e100dc  absl::logging_internal::LogMessage::DieIfFatal()
    @     0x55c202e0eb56  absl::logging_internal::LogMessage::SendToLog()
    @     0x55c202e0e441  absl::logging_internal::LogMessage::Flush()
    @     0x55c202e10c89  absl::logging_internal::LogMessageFatal::~LogMessageFatal()
    @     0x55c1ff2eb768  xla::ShapeUtil::ElementsIn()
    @     0x55c1fee0e81d  xla::HloDataflowAnalysis::CanShareOperandBufferWithUser()
    @     0x55c1fedc4147  xla::HeapSimulator::RunComputation()
    @     0x55c1fedc297f  xla::HeapSimulator::Run()
    @     0x55c1fedc253c  xla::HeapSimulator::MinimumMemoryForComputation()
    @     0x55c1fa3d72a7  xla::ListMemoryScheduler()
    @     0x55c1fa3da29e  xla::DefaultMemoryScheduler()
    @     0x55c1fa3dbeb9  xla::(anonymous namespace)::ScheduleComputationHelper()
    @     0x55c1fa3dbc93  xla::ScheduleComputation()
    @     0x55c1f5349ffb  xla::gpu::GpuHloSchedule::Build()
    @     0x55c1f5318574  xla::gpu::CompileModuleToLlvmIrImpl()
    @     0x55c1f5317348  xla::gpu::GpuCompiler::RunBackend()
    @     0x55c1fc36ca27  xla::Service::BuildExecutable()
    @     0x55c1fc3630f0  xla::LocalService::CompileExecutables()
    @     0x55c1fc206c1e  xla::LocalClient::Compile()
    @     0x55c1fc1ebd6b  xla::PjRtStreamExecutorClient::Compile()
    @     0x55c1f8f25c83  xla::PyClient::Compile()
```

Do you think that's enough ? To repro you could install JAX (it's completely OSS), run that test, capture the HLO, and observe it.

trentlo commented:
> ElementsIn

Thanks for the info. Let me try reproducing it by installing JAX. Will ask you for more if that is not enough.

Looks like it might be some misuse of ElementsIn() according to the dump.

cheshire commented:
Well it's a CHECK failure inside XLA, so it's definitely a bug.

akuegel commented:
Yes, based on the stack trace, this is a call to the ElementsIn() function on a tuple shape. The PR adds three lines which call ElementsIn(). The call in ConcatIsEffectivelyElementwise should be fine, because the operand of a concat cannot have a tuple shape. This leaves the two calls in CanShareOperandBufferWithUser, which seems to also match what we see in the stack trace. I guess you need to add a check to the condition for whether it is a tuple shape or not.

trentlo commented:
> Yes, based on the stack trace, this is a call to the ElementsIn() function on a tuple shape. The PR adds three lines which call ElementsIn(). The call in ConcatIsEffectivelyElementwise should be fine, because the operand of a concat cannot have a tuple shape. This leaves the two calls in CanShareOperandBufferWithUser, which seems to also match what we see in the stack trace. I guess you need to add a check to the condition for whether it is a tuple shape or not.

Right. That is what I had in mind. Thanks for the analysis!

trentlo commented:
> https://cs.opensource.google/jax/jax/+/master:tests/jet_test.py;l=297?q=JetTest.test_cumsum&sq=

@cheshire I dumped the HLOs from cumsum in https://cs.opensource.google/jax/jax/+/master:tests/jet_test.py;l=297?q=JetTest.test_cumsum&sq= but I do not repro the crash with xla replay.

Is it possible that the crashed test is another?

trentlo commented:
> > https://cs.opensource.google/jax/jax/+/master:tests/jet_test.py;l=297?q=JetTest.test_cumsum&sq=
> 
> @cheshire I dumped the HLOs from cumsum in https://cs.opensource.google/jax/jax/+/master:tests/jet_test.py;l=297?q=JetTest.test_cumsum&sq= but I do not repro the crash with xla replay.
> 
> Is it possible that the crashed test is another?

@cheshire I meant if you could double check whether the failing test is cumsum? Or I just should proceed to build JAX and see if I can repro the bug with running JAX?

cheshire commented:
@trentlo sorry for the inconvenience! The test failure was in that file. You can also try to run all JAX tests with your patch:

```
git clone https://github.com/tensorflow/tensorflow.git
# check out branch
git clone https://github.com/google/jax.git
cd jax
python build/build.py --bazel_options="--override_repository=org_tensorflow=path/to/checked/out/tensorflow"
pip install dist/*.whl
pip install -e .
pytest -n auto tests examples
```

trentlo commented:
> @trentlo sorry for the inconvenience! The test failure was in that file. You can also try to run all JAX tests with your patch:
> 
> ```
> git clone https://github.com/tensorflow/tensorflow.git
> # check out branch
> git clone https://github.com/google/jax.git
> cd jax
> python build/build.py --bazel_options="--override_repository=org_tensorflow=path/to/checked/out/tensorflow"
> pip install dist/*.whl
> pip install -e .
> pytest -n auto tests examples
> ```

I see. Thanks for the reply. Will try to repro with building JAX.

trentlo commented:
> > @trentlo sorry for the inconvenience! The test failure was in that file. You can also try to run all JAX tests with your patch:
> > ```
> > git clone https://github.com/tensorflow/tensorflow.git
> > # check out branch
> > git clone https://github.com/google/jax.git
> > cd jax
> > python build/build.py --bazel_options="--override_repository=org_tensorflow=path/to/checked/out/tensorflow"
> > pip install dist/*.whl
> > pip install -e .
> > pytest -n auto tests examples
> > ```
> 
> I see. Thanks for the reply. Will try to repro with building JAX.

@cheshire I built JAX with overriding the tensorflow path as you said and tested with `pytest -n 1 tests/jet_test.py`. However, I don't repro the issue as the stack dump you showed (i.e., the test runs through successfully). (BTW I'm quite sure the tensorflow source I provided is hooked.)

I am trying running though all the JAX tests. But are you sure the failure is from jet_test.py?

BTW, I saw some CUDA_ERROR_OUT_OF_MEMORY when I ran some tests other than jet_test.py. How could I limit the memory usage for the tests? I run on 16GB memory.

trentlo commented:
> > > @trentlo sorry for the inconvenience! The test failure was in that file. You can also try to run all JAX tests with your patch:
> > > ```
> > > git clone https://github.com/tensorflow/tensorflow.git
> > > # check out branch
> > > git clone https://github.com/google/jax.git
> > > cd jax
> > > python build/build.py --bazel_options="--override_repository=org_tensorflow=path/to/checked/out/tensorflow"
> > > pip install dist/*.whl
> > > pip install -e .
> > > pytest -n auto tests examples
> > > ```
> > 
> > 
> > I see. Thanks for the reply. Will try to repro with building JAX.
> 
> @cheshire I built JAX with overriding the tensorflow path as you said and tested with `pytest -n 1 tests/jet_test.py`. However, I don't repro the issue as the stack dump you showed (i.e., the test runs through successfully). (BTW I'm quite sure the tensorflow source I provided is hooked.)
> 
> I am trying running though all the JAX tests. But are you sure the failure is from jet_test.py?
> 
> BTW, I saw some CUDA_ERROR_OUT_OF_MEMORY when I ran some tests other than jet_test.py. How could I limit the memory usage for the tests? I run on 16GB memory.

Update:
Ran through all the tests. See a bunch of CUDA_ERROR_OUT_OF_MEMORY. Do not see other failures.

cheshire commented:
OK I'll run the tests internally myself and try to figure out where the
failures are coming from!

On Tue, Feb 23, 2021 at 6:29 PM Trent Lo <notifications@github.com> wrote:

> @trentlo <https://github.com/trentlo> sorry for the inconvenience! The
> test failure was in that file. You can also try to run all JAX tests with
> your patch:
>
> git clone https://github.com/tensorflow/tensorflow.git
> # check out branch
> git clone https://github.com/google/jax.git
> cd jax
> python build/build.py --bazel_options="--override_repository=org_tensorflow=path/to/checked/out/tensorflow"
> pip install dist/*.whl
> pip install -e .
> pytest -n auto tests examples
>
> I see. Thanks for the reply. Will try to repro with building JAX.
>
> @cheshire <https://github.com/cheshire> I built JAX with overriding the
> tensorflow path as you said and tested with pytest -n 1 tests/jet_test.py.
> However, I don't repro the issue as the stack dump you showed (i.e., the
> test runs through successfully). (BTW I'm quite sure the tensorflow source
> I provided is hooked.)
>
> I am trying running though all the JAX tests. But are you sure the failure
> is from jet_test.py?
>
> BTW, I saw some CUDA_ERROR_OUT_OF_MEMORY when I ran some tests other than
> jet_test.py. How could I limit the memory usage for the tests? I run on
> 16GB memory.
>
> Update:
> Ran through all the tests. See a bunch of CUDA_ERROR_OUT_OF_MEMORY. Do not
> see other failures.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/46614#issuecomment-784702227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACVGHYDULD47RXMTKL6UCTTARP7DANCNFSM4WO7K7ZQ>
> .
>

cheshire commented:
Are you running the debug build ? The specified test passes with `-c opt`, but fails with `-c dbg`.

cheshire commented:
The problem reproduces on this HLO in debug build: https://gist.github.com/cheshire/765c08aae5a3d20ee00be65aeabc43db

trentlo commented:
Ah. That makes sense. I ran without `-c dbg`. Thanks!

> Are you running the debug build ? The specified test passes with `-c opt`, but fails with `-c dbg`.

trentlo commented:
Strange. I built the replay_computation with the below command but still don't repro the bug.

`bazel build -c dbg --config=cuda --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --output_filter MATCH_NOTHING --disk_cache=/mnt/trentl/bazel_cache.upstream-master/ //tensorflow/compiler/xla/tools:replay_computation_gpu`

Run with the following command:
`./bazel-bin/tensorflow/compiler/xla/tools/replay_computation_gpu --use_fake_data=true --print_result=false --compile_only=false george.hlo`

trentlo commented:
You meant you ran replay_computation or JAX? What is the command you used to build and run tests?

> Are you running the debug build ? The specified test passes with `-c opt`, but fails with `-c dbg`.

cheshire commented:
@trentlo could you try

```
tensorflow/compiler/xla/tools:run_hlo_module -- --platform=CUDA
```

?

cheshire commented:
also turn these two lines into `CHECK` and `CHECK_EQ`: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/shape_util.cc;l=536-537?q=file:shape_util.cc

cheshire commented:
The actual error is: `Check failed: shape.IsArray() (f64[6], f64[6], f64[6], f64[6])`

trentlo commented:
Did what you suggested but still no repro. It is good that you confirm that we do see a tuple shape in `Check failed: shape.IsArray()`. This makes sense and it is aligned to the stack dump at the very beginning of the thread. The only explanation to the difference we see might be some code diff between OSS and your internal TF/XLA codebase? Or there could be some configuration differences (e.g., some optimizations are by default off in OSS while on in your internal BUILD)?

Another way to debug may be to provide me all the hlos from the testcase jit__cumulative_reduction.115 with `--xla_dump_hlo_pass_re=.*`. What do you think?
 
Change the DCHECK to CHECK as suggested and observe the below result. Also try with JAX. BTW, I'm sure my TF is the latest revision from the OSS repro.

```
xxx@xxx:/opt/tensorflow/tensorflow-source# ./bazel-bin/tensorflow/compiler/xla/tools/run_hlo_module --platform=cuda george.hlo
2021-02-24 20:39:51.354027: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.10.2
2021-02-24 20:39:51.411459: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
Running HLO module on platform CUDA...
... compiled and ran in 1.0083s.
Running HLO module on platform Interpreter...
... compiled and ran in 0.033454s.

** Results on cuda and Interpreter are close enough. **
0/1 runs miscompared.

```

trentlo commented:
Ah~ @cheshire, it is my bad. I just realized that I merged the upstream TF so that I also merged the `revert` commit... (It canceled out the (`Make H-loop-fusion share operands with users`) changes in my local change and the upstream master and my local branch are merged perfectly) That is why I can't repro the bug. I repro it now.

Thanks for all the help. Sorry for the bothering.
